src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -1407,6 +1434,74 @@ public ValueTypedFetchProperty(Type type, PropertyInfo property) : base(type)
                         private readonly StructFunc<TStruct, TProperty> _propertyFetch;
                     }
 
+#if NETCOREAPP
+                    /// <summary>
+                    /// A fetcher that can be used when MakeGenericType isn't available.
+                    /// </summary>
+                    private sealed class ReflectionPropertyFetch : PropertyFetch
+                    {
+                        private readonly PropertyInfo _property;
+                        public ReflectionPropertyFetch(Type type, PropertyInfo property) : base(type)
+                        {
+                            _property = property;
+                        }
+
+                        public override object? Fetch(object? obj) => _property.GetValue(obj);
+                    }
+
+                    /// <summary>
+                    /// A fetcher that enumerates and formats an IEnumerable when MakeGenericType isn't available.
+                    /// </summary>
+                    private sealed class EnumeratePropertyFetch : PropertyFetch
+                    {
+                        public EnumeratePropertyFetch(Type type) : base(type) { }
+
+                        public override object? Fetch(object? obj)
+                        {
+                            IEnumerable? enumerable = obj as IEnumerable;
+                            Debug.Assert(enumerable is not null);
+
+                            // string.Join for a non-generic IEnumerable
+                            IEnumerator en = enumerable.GetEnumerator();
+                            using (IDisposable? disposable = en as IDisposable)
+                            {
+                                if (!en.MoveNext())
+                                {
+                                    return string.Empty;
+                                }
+
+                                object? currentValue = en.Current;
+                                string? firstString = currentValue?.ToString();
+
+                                // If there's only 1 item, simply return the ToString of that
+                                if (!en.MoveNext())
+                                {
+                                    // Only one value available
+                                    return firstString ?? string.Empty;
+                                }
+
+                                var result = new ValueStringBuilder(stackalloc char[256]);
+
+                                result.Append(firstString);
```

Comment:
firstString.Length can be greater than 256?

Reply:
ValueStringBuilder handles growing and allocating as necessary.